lite/src/types.rs

```diff
@@ -0,0 +1,103 @@
+/// TrustedState stores the latest state trusted by a lite client,
+/// including the last header and the validator set to use to verify
+/// the next header.
+pub struct TrustedState<H, V>
+where
+    H: Header,
+    V: Validators,
+{
+    pub header: H,
+    pub next_validators: V,
+}
+
+/// Need to do something better here :)
+pub type Height = u64;
+pub type Hash = u64; // TODO
+pub type Time = u64; // TODO
+pub type Bytes = u64; // TODO
+pub type ValID = u64; // TODO
+
+/// Header contains meta data about the block -
+/// the height, the time, the hash of the validator set
+/// that should sign this header, and the hash of the validator
+/// set that should sign the next header.
+pub trait Header {
+    fn height(&self) -> Height;
+    fn bft_time(&self) -> Time;
+    fn validators_hash(&self) -> Hash;
+    fn next_validators_hash(&self) -> Hash;
+
+    /// Hash of the header (ie. the hash of the block).
+    fn hash(&self) -> Hash;
+}
+
+/// Validators is the full validator set.
+/// It exposes its hash, which should match whats in a header,
+/// and its total power. It also has an underlying
+/// Validator type which can be used for verifying signatures.
+pub trait Validators {
+    type Validator: Validator;
+
+    /// Hash of the validator set.
+    fn hash(&self) -> Hash;
+
+    /// Total voting power of the set
+    fn total_power(&self) -> u64;
+
+    /// For iterating over the underlying validators.
+    fn into_vec(&self) -> Vec<Self::Validator>;
```

Comment:
Minor detail: it seems we are always using it as iterator so that might be more appropriate type, but maybe vector makes more sense for future extensions.

Reply:
You could probably use `impl Iterator<Item=&Self::Validator>` as the return type here, then call `iter()` in an implementation

Reply:
Can you provide a bit more detail here, not sure I fully understand.

If I do `fn into_iter(&self) -> Iterator<Item = &Self::Validator>;` I get 

```
the size for values of type `dyn std::iter::Iterator<Item = &<V as types::Validators>::Validator>` cannot be known at compilation time
```

Also `impl` doesn't seem to be allowed directly in the return type? Ie.

```
error[E0562]: `impl Trait` not allowed outside of function and inherent method return types
  --> lite/src/types.rs:48:28
   |
48 |     fn into_iter(&self) -> impl Iterator<Item = Self::Validator>;
```

Reply:
Aah, my bad, guess you can't use `impl Trait` in a trait definition, in which case you'd need a concrete iterator type.

In that case, you could potentially use an associated type e.g. `type ValidatorIter`.

One note about `into_iter` as a name: that's the name used by the [`IntoIterator` trait](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html), which is for iterators that consume `self`. The convention for a borrowing iterator is `fn iter(&self) -> MyIter`

Reply:
Wow that worked! Thanks! https://github.com/interchainio/tendermint-rs/pull/31/commits/f521f048d6a25b4ffd0efcf2c788ba17d7dcf95d

cc @spacejam

Reply:
Eek this is probably going to over complicate implementations - I haven't figured out yet how a concrete type (eg. the `validator::Set`) can actually implement this without a trait object. It seems we have to implement ExactSizedIterator for it, but haven't made that work easily yet.